In [30]:
import pandas as pd
import re
import nltk
from nltk.stem.porter import PorterStemmer
from nltk.corpus import stopwords
from bs4 import BeautifulSoup
from datasets import load_dataset


In [31]:
data = pd.read_csv(r"C:\Users\srika\OneDrive\Documents\York\Sem-2 york\MBAN 6090 - Analytics Consulting Project\webscrape\communitech\jobs_info.csv")

In [32]:
data.head()

Unnamed: 0 organization_name  \
0           1           Vidyard   
1           2         Hootsuite   
2           3            Semios   
3           4           Axonify   
4           5     Deep Genomics   

                                               title  \
0          Director of Sales, Enterprise Acquisition   
1  Senior Software Developer | Développeur logici...   
2                                    Chemistry Co-op   
3  Senior Talent Acquisition Partner (15 months c...   
4  (Senior) Research Scientist - Statistical Gene...   

                                          locations  \
0             ['United States', 'Canada', 'Remote']   
1  ['Toronto, ON, Canada', 'Vancouver, BC, Canada']   
2                         ['Vancouver, BC, Canada']   
3   ['Toronto, ON, Canada', 'Waterloo, ON, Canada']   
4                           ['Toronto, ON, Canada']   

                                 organization_topics  \
0  ['B2B', 'Media & entertainment', 'Productivity...   
1  ['500+', 'Advertising/marketing', 'B2B', 'Soft...   
2  ['Agriculture', 'B2B', 'C100 alumni', 'Data & ...   
3  ['B2B', 'Data & analytics', 'Education', 'Fema...   
4  ['50-99', 'AI/ML', 'B2B', 'Biotechnology', 'Me...   

                          organization_industry_tags  \
0                              ['Media', 'Software']   
1  ['Information Technology', 'Media', 'Sales and...   
2                        ['Agriculture and Farming']   
3  ['Education', 'Information Technology', 'Softw...   
4  ['Biotechnology', 'DeepTech', 'Health', 'Softw...   

                                                 url  has_description  \
0  https://www1.communitech.ca/companies/vidyard/...             True   
1  https://www1.communitech.ca/companies/hootsuit...             True   
2  https://www1.communitech.ca/companies/semios/j...             True   
3  https://www1.communitech.ca/companies/axonify/...             True   
4  https://www1.communitech.ca/companies/deep-gen...             True   

                                         description  
0  <div data-testid="careerPage"><p>At Vidyard, w...  
1  <div data-testid="careerPage"><p><span><em><st...  
2  <div data-testid="careerPage"><p><strong>Who w...  
3  <div data-testid="careerPage"><section><div>Ax...  
4  <div data-testid="careerPage"><section><div><b...

In [33]:
data.iloc[123]['description']

'<div data-testid="careerPage"><p><strong>Your role as In-person Network Coordinator</strong></p><p>Reporting to the EAP Manager, the In-person Network Coordinator plays a vital role in enhancing our provider network\'s reach and effectiveness, as well as ensuring comprehensive support to our external mental health providers.</p><p><strong>What you’ll be doing: </strong></p><ul> <li>Execution of in-person provider recruitment, such as candidate sourcing, inbox management and contract drafting, in close collaboration with the Talent Acquisition Specialist</li> <li>Support manager in tracking and planning network coverage across Canada</li> <li>Provide seamless onboarding and off-boarding experience, keeping internal management tools up-to-date and offering personalized coaching to providers to drive completion of HR and IT requirements</li> <li>Provide and maintain a great ongoing provider experience, with responsive support and answer inquiries from providers, addressing their question

In [34]:
# keywords = {
#     "Core Responsibilities": ["responsibilities", "role", "duties"],
#     "Required Skills": ["skills", "qualifications", "requirements"],
#     "Preferred Qualifications": ["preferred", "plus", "advantage"],
#     "Compensation and Benefits": ["compensation", "benefits", "salary"]
# }

# def process_description(description):
#     if not isinstance(description, str):
#         return {key: "N/A" for key in keywords}

#     soup = BeautifulSoup(description, 'html.parser')
#     text = soup.get_text(separator=' ').lower()

#     job_info = {key: "N/A" for key in keywords}

#     for key, key_list in keywords.items():
#         for phrase in key_list:
#             if phrase in text:
#                 start_idx = text.find(phrase)
#                 end_idx = text.find('.', start_idx) + 1
#                 job_info[key] = text[start_idx:end_idx].strip()
#                 break

#     job_info["Educational Requirements"] = "N/A"
#     job_info["Experience Level"] = "N/A"

#     return job_info


# description = data.iloc[123]['description']
# processed_description = process_description(description)


In [35]:
# processed_descriptions = []  # List to hold processed descriptions

# for job in data['description']:
#     processed_description = process_description(job)
#     processed_descriptions.append(processed_description)

# data_2 = pd.DataFrame(processed_descriptions)
# data_2.head()


In [36]:
dataset = load_dataset("jacob-hugging-face/job-descriptions")


print(dataset)


DatasetDict({
    train: Dataset({
        features: ['company_name', 'job_description', 'position_title', 'description_length', 'model_response'],
        num_rows: 853
    })
})


In [37]:
data_3 = pd.DataFrame(dataset)

In [38]:
data_3.head()

train
0  {'company_name': 'Google', 'job_description': ...
1  {'company_name': 'Apple', 'job_description': '...
2  {'company_name': 'Netflix', 'job_description':...
3  {'company_name': 'Robert Half', 'job_descripti...
4  {'company_name': 'TrackFive', 'job_description...

In [39]:
data.head()

Unnamed: 0 organization_name  \
0           1           Vidyard   
1           2         Hootsuite   
2           3            Semios   
3           4           Axonify   
4           5     Deep Genomics   

                                               title  \
0          Director of Sales, Enterprise Acquisition   
1  Senior Software Developer | Développeur logici...   
2                                    Chemistry Co-op   
3  Senior Talent Acquisition Partner (15 months c...   
4  (Senior) Research Scientist - Statistical Gene...   

                                          locations  \
0             ['United States', 'Canada', 'Remote']   
1  ['Toronto, ON, Canada', 'Vancouver, BC, Canada']   
2                         ['Vancouver, BC, Canada']   
3   ['Toronto, ON, Canada', 'Waterloo, ON, Canada']   
4                           ['Toronto, ON, Canada']   

                                 organization_topics  \
0  ['B2B', 'Media & entertainment', 'Productivity...   
1  ['500+', 'Advertising/marketing', 'B2B', 'Soft...   
2  ['Agriculture', 'B2B', 'C100 alumni', 'Data & ...   
3  ['B2B', 'Data & analytics', 'Education', 'Fema...   
4  ['50-99', 'AI/ML', 'B2B', 'Biotechnology', 'Me...   

                          organization_industry_tags  \
0                              ['Media', 'Software']   
1  ['Information Technology', 'Media', 'Sales and...   
2                        ['Agriculture and Farming']   
3  ['Education', 'Information Technology', 'Softw...   
4  ['Biotechnology', 'DeepTech', 'Health', 'Softw...   

                                                 url  has_description  \
0  https://www1.communitech.ca/companies/vidyard/...             True   
1  https://www1.communitech.ca/companies/hootsuit...             True   
2  https://www1.communitech.ca/companies/semios/j...             True   
3  https://www1.communitech.ca/companies/axonify/...             True   
4  https://www1.communitech.ca/companies/deep-gen...             True   

                                         description  
0  <div data-testid="careerPage"><p>At Vidyard, w...  
1  <div data-testid="careerPage"><p><span><em><st...  
2  <div data-testid="careerPage"><p><strong>Who w...  
3  <div data-testid="careerPage"><section><div>Ax...  
4  <div data-testid="careerPage"><section><div><b...

In [40]:
ps = PorterStemmer()

In [41]:
def cleaning(txt):
    cleaned_txt = re.sub(r'[^a-zA-Z0-9\s]', ' ', txt)
    token = nltk.word_tokenize(cleaned_txt)
    return token


In [42]:
data.iloc[12]['description']

'<div data-testid="careerPage"><section><div>At Horizon, we make web3 easy, fun, and accessible for users and builders. Our mission is to unlock web3\'s utility, drive mass adoption, and accelerate content creation by empowering builders to create amazing web3 games and experiences.</div> <div>We have two powerful products to achieve our mission:</div> <div>* Sequence, the leading all-in-one developer platform for building web3 games and apps.</div> <div>* Skyweaver, a digital trading card game—powered by Sequence—where players can win, own, and trade their NFT/SFT game items.</div> <div>Sequence unlocks web3’s utility, and is poised to accelerate web3 content creation and drive adoption on a massive scale. Skyweaver—named 2022’s best blockchain game—is a groundbreaking video game with an unprecedented player-owned economy, novel gameplay, and passionate community.</div> <div>While the challenge is hard, we find it incredibly fun. And, we have the ingredients to make it happen: a well-

In [43]:
cleaning(data.iloc[123]['description'])

['div',
 'data',
 'testid',
 'careerPage',
 'p',
 'strong',
 'Your',
 'role',
 'as',
 'In',
 'person',
 'Network',
 'Coordinator',
 'strong',
 'p',
 'p',
 'Reporting',
 'to',
 'the',
 'EAP',
 'Manager',
 'the',
 'In',
 'person',
 'Network',
 'Coordinator',
 'plays',
 'a',
 'vital',
 'role',
 'in',
 'enhancing',
 'our',
 'provider',
 'network',
 's',
 'reach',
 'and',
 'effectiveness',
 'as',
 'well',
 'as',
 'ensuring',
 'comprehensive',
 'support',
 'to',
 'our',
 'external',
 'mental',
 'health',
 'providers',
 'p',
 'p',
 'strong',
 'What',
 'you',
 'll',
 'be',
 'doing',
 'strong',
 'p',
 'ul',
 'li',
 'Execution',
 'of',
 'in',
 'person',
 'provider',
 'recruitment',
 'such',
 'as',
 'candidate',
 'sourcing',
 'inbox',
 'management',
 'and',
 'contract',
 'drafting',
 'in',
 'close',
 'collaboration',
 'with',
 'the',
 'Talent',
 'Acquisition',
 'Specialist',
 'li',
 'li',
 'Support',
 'manager',
 'in',
 'tracking',
 'and',
 'planning',
 'network',
 'coverage',
 'across',
 'Canada

In [44]:
data = pd.read_csv(r"C:\Users\srika\OneDrive\Documents\York\Sem-2 york\MBAN 6090 - Analytics Consulting Project\webscrape\communitech\jobs_info.csv")

unique_tags = {}  

for description in data['description']:
    if not isinstance(description, str):
        continue

    soup = BeautifulSoup(description, 'html.parser')
    for tag in soup.find_all(True):
        tag_name = tag.name
        if tag_name in unique_tags:
            unique_tags[tag_name] += 1
        else:
            unique_tags[tag_name] = 1

unique_tags


{'div': 50342,
 'p': 45547,
 'h2': 2277,
 'strong': 16978,
 'ul': 13861,
 'li': 91025,
 'em': 1656,
 'span': 74934,
 'hr': 186,
 'a': 12531,
 'br': 8977,
 'section': 5100,
 'h3': 4353,
 'b': 11358,
 'u': 654,
 'i': 1618,
 'sub': 20,
 'h6': 22,
 'ol': 43,
 'h4': 673,
 'h1': 377,
 'header': 11,
 'img': 726,
 'nav': 169,
 'button': 156,
 'main': 10,
 'footer': 87,
 'dl': 3,
 'dt': 52,
 'dd': 52,
 'svg': 102,
 'g': 4,
 'path': 57,
 'style': 123,
 'table': 37,
 'tbody': 37,
 'tr': 108,
 'td': 203,
 'h5': 145,
 's': 2,
 'form': 88,
 'input': 723,
 'label': 142,
 'fieldset': 8,
 'legend': 8,
 'textarea': 8,
 'abbr': 24,
 'select': 26,
 'option': 124,
 'blockquote': 5,
 'sup': 7,
 'font': 323,
 'time': 202,
 'thead': 3,
 'th': 9,
 'article': 1,
 'aside': 2,
 'circle': 2,
 'small': 1}

In [45]:
unique_tags_df = pd.DataFrame(list(unique_tags.items()), columns=['Tag', 'Count'])

unique_tags_df = unique_tags_df.sort_values(by='Count', ascending=False)
unique_tags_df.reset_index(drop=True, inplace=True)


unique_tags_df.head(10)


Tag  Count
0       li  91025
1     span  74934
2      div  50342
3        p  45547
4   strong  16978
5       ul  13861
6        a  12531
7        b  11358
8       br   8977
9  section   5100

In [46]:
head = unique_tags_df[unique_tags_df['Tag'].isin(['h1', 'h2', 'h3', 'h4', 'h5', 'h6'])]

In [47]:
unique_tags_df = unique_tags_df.head(20)
unique_tags_df = unique_tags_df[~unique_tags_df['Tag'].isin(['span','strong', 'div', 'img', 'input', 'font', 'u','em','i','a'])]


In [48]:
unique_tags_df = pd.merge(unique_tags_df, head, on='Tag', how='outer')
unique_tags_df.drop_duplicates(inplace=True)
unique_tags_df['count'] = unique_tags_df['Count_x'].fillna(unique_tags_df['Count_y'])
unique_tags_df.drop(['Count_x', 'Count_y'], axis=1, inplace=True)
unique_tags_df



Tag    count
0        li  91025.0
1         p  45547.0
2        ul  13861.0
3         b  11358.0
4        br   8977.0
5   section   5100.0
6        h3   4353.0
7        h2   2277.0
8        h4    673.0
9        h1    377.0
10       h5    145.0
11       h6     22.0

In [49]:
li_data = [] 

for description in data['description']:
    if isinstance(description, str):
        soup = BeautifulSoup(description, 'html.parser')
    
        def find_li_tags(tag, level):
            if tag.name == 'li':
                li_text = tag.get_text()
                li_data.append({'text': li_text, 'level': level})
        
            for child in tag.children:
                if child.name:
                    find_li_tags(child, level + 1)
    
        find_li_tags(soup, level=1)

tag_descriptions = pd.DataFrame(li_data)

tag_descriptions.head(10)

text  level
0  Driving a culture of achievement to exceed rev...      4
1  Setting clear expectations around revenue gene...      4
2  Working with marketing for in-market events an...      4
3  Coaching team in effective sales strategies, p...      4
4  Cultivating and maintaining strong relationshi...      4
5  Regularly analyzing sales metrics, pipeline da...      4
6  Continuously refining and enhancing the sales ...      4
7  ~ 5 years of proven sales leadership experienc...      4
8  Proficiency in analyzing sales data and metric...      4
9  Experience managing and closing complex B2B sa...      4

In [50]:
tag_descriptions['level'].value_counts()

level
4     38139
5     32789
6      8384
7      5473
11     3149
9       955
8       904
13      549
10      249
3       247
12       58
21       32
18       28
14       25
16       15
15       15
20       14
Name: count, dtype: int64

In [51]:
tag_descriptions.loc[tag_descriptions['level'] == 20]['text']

60276    Writing automated functional and performance t...
60277    Troubleshooting system and network problems an...
60278    Reproducing issues in the lab and testing fixe...
60279    Writing procedural documentation and event rep...
60282    Academic and ideally professional experience w...
60283               Experience with Python, BASH, and REST
60284    Knowledge of some of the following areas and d...
60285         GIT, bug tracking software and documentation
60286    Excellent problem solving and troubleshooting ...
60287        Ability to write clear procedures and reports
60288    Ability to perform under pressure in a deadlin...
60289    Excellent team player with a high level of sel...
60290    Excellent communication skills, both verbal an...
60291    B Sc. in Computer Science, Software Engineerin...
Name: text, dtype: object

In [52]:
first_line = tag_descriptions.loc[tag_descriptions['level'] == 20]['text'].iloc[0]

matching_description = data[data['description'].fillna('').str.contains(first_line)]['description'].iloc[0]

matching_description



'<div data-testid="careerPage"><div class="so-panel widget widget_sow-editor panel-first-child" data-index="4" id="panel-18-1-0-0">\n<div class="marbot-20 panel-widget-style panel-widget-style-for-18-1-0-0">\n<div class="so-widget-sow-editor so-widget-sow-editor-base">\n<div class="siteorigin-widget-tinymce textwidget">\n<div class="so-panel widget widget_sow-editor panel-first-child" data-index="4" id="panel-18-1-0-0">\n<div class="marbot-20 panel-widget-style panel-widget-style-for-18-1-0-0">\n<div class="so-widget-sow-editor so-widget-sow-editor-base">\n<div class="siteorigin-widget-tinymce textwidget">\n<div class="so-panel widget widget_sow-editor panel-first-child" data-index="6" id="panel-18-2-0-0">\n<div class="marbot-20 panel-widget-style panel-widget-style-for-18-2-0-0">\n<div class="so-widget-sow-editor so-widget-sow-editor-base">\n<div class="siteorigin-widget-tinymce textwidget">\n<h4>Test Automation Software Developer</h4>\n<h6><span style="font-weight: 400;">Job ID: 2023

In [58]:
<div data-testid="careerPage"><div class="so-panel widget widget_sow-editor panel-first-child" data-index="4" id="panel-18-1-0-0">\n<div class="marbot-20 panel-widget-style panel-widget-style-for-18-1-0-0">\n<div class="so-widget-sow-editor so-widget-sow-editor-base">\n<div class="siteorigin-widget-tinymce textwidget">\n<div class="so-panel widget widget_sow-editor panel-first-child" data-index="4" id="panel-18-1-0-0">\n<div class="marbot-20 panel-widget-style panel-widget-style-for-18-1-0-0">\n<div class="so-widget-sow-editor so-widget-sow-editor-base">\n<div class="siteorigin-widget-tinymce textwidget">\n<div class="so-panel widget widget_sow-editor panel-first-child" data-index="6" id="panel-18-2-0-0">\n<div class="marbot-20 panel-widget-style panel-widget-style-for-18-2-0-0">\n<div class="so-widget-sow-editor so-widget-sow-editor-base">\n<div class="siteorigin-widget-tinymce textwidget">\n<h4>Test Automation Software Developer</h4>\n<h6><span style="font-weight: 400;">Job ID: 2023070104</span></h6>\n<p><b>Job Description</b></p>\n<p><span style="font-weight: 400;">You will design and develop an automated test framework designed for application layer and embedded software for bleeding edge networking technologies including ultra-fast network processors (up to 12.8Tbps). The products bring together Open Systems, Network Virtualization and fully Programmable Network Logic to meet the needs of Data Centers, Network Service Providers and researchers in Software Defined Networking technology.</span></p>\n<p>\xa0</p>\n<p><strong>Responsibilities</strong></p>\n<ul>\n<li style="list-style-type: none;">\n<ul>\n<li style="list-style-type: none;">\n<ul>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Writing automated functional and performance test cases from requirements, executing them in the lab, troubleshooting issues to identify their root cause and reporting results</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Troubleshooting system and network problems and diagnosing and solving hardware or software faults</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Reproducing issues in the lab and testing fixes and workarounds</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Writing procedural documentation and event reports</span></li>\n</ul>\n</li>\n</ul>\n</li>\n</ul>\n<p>\xa0</p>\n<p><strong>Qualifications and Skills Required</strong></p>\n<ul>\n<li style="list-style-type: none;">\n<ul>\n<li style="list-style-type: none;">\n<ul>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Academic and ideally professional experience writing software and/or scripts</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Experience with Python, BASH, and REST</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Knowledge of some of the following areas and drive to learn more: L2 to L4 of TCP/IP, traffic generators, hardware interfaces, chipset based solutions.</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">GIT, bug tracking software and documentation</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Excellent problem solving and troubleshooting skills</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Ability to write clear procedures and reports</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Ability to perform under pressure in a deadline driven environment</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Excellent team player with a high level of self-motivation and initiative</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Excellent communication skills, both verbal and written (multilingualism is an asset)</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">B Sc. in Computer Science, Software Engineering, B.Elec.Eng. completed</span></li>\n</ul>\n</li>\n</ul>\n</li>\n</ul>\n<p>\xa0</p>\n<p><b>Good to have</b></p>\n<ul>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Networking certifications</span></li>\n</ul>\n</div>\n</div>\n</div>\n</div>\n<div class="so-panel widget widget_sow-editor panel-last-child" data-index="7" id="panel-18-2-0-1">\n<div class="so-widget-sow-editor so-widget-sow-editor-base">\n<div class="siteorigin-widget-tinymce textwidget">\n<p>\xa0</p>\n<p><strong>Additional Information</strong></p>\n<p>Type: Full-time</p>\n<p>Location: Montreal, QC, Canada</p>\n<hr/>\n<p>For more\xa0information, or to submit your resumé, please\xa0e-mail\xa0NoviFlow at\xa0<a href="https://mail.google.com/mail/?view=cm&amp;fs=1&amp;tf=1&amp;to=careers@noviflow.com">careers@noviflow.com</a></p>\n</div>\n</div>\n</div>\n</div>\n</div>\n</div>\n</div>\n</div>\n</div>\n</div>\n</div></div>'


SyntaxError: invalid syntax (930925038.py, line 1)

<div data-testid="careerPage"><div class="so-panel widget widget_sow-editor panel-first-child" data-index="4" id="panel-18-1-0-0">\n<div class="marbot-20 panel-widget-style panel-widget-style-for-18-1-0-0">\n<div class="so-widget-sow-editor so-widget-sow-editor-base">\n<div class="siteorigin-widget-tinymce textwidget">\n<div class="so-panel widget widget_sow-editor panel-first-child" data-index="4" id="panel-18-1-0-0">\n<div class="marbot-20 panel-widget-style panel-widget-style-for-18-1-0-0">\n<div class="so-widget-sow-editor so-widget-sow-editor-base">\n<div class="siteorigin-widget-tinymce textwidget">\n<div class="so-panel widget widget_sow-editor panel-first-child" data-index="6" id="panel-18-2-0-0">\n<div class="marbot-20 panel-widget-style panel-widget-style-for-18-2-0-0">\n<div class="so-widget-sow-editor so-widget-sow-editor-base">\n<div class="siteorigin-widget-tinymce textwidget">\n<h4>Test Automation Software Developer</h4>\n<h6><span style="font-weight: 400;">Job ID: 2023070104</span></h6>\n<p><b>Job Description</b></p>\n<p><span style="font-weight: 400;">You will design and develop an automated test framework designed for application layer and embedded software for bleeding edge networking technologies including ultra-fast network processors (up to 12.8Tbps). The products bring together Open Systems, Network Virtualization and fully Programmable Network Logic to meet the needs of Data Centers, Network Service Providers and researchers in Software Defined Networking technology.</span></p>\n<p>\xa0</p>\n<p><strong>Responsibilities</strong></p>\n<ul>\n<li style="list-style-type: none;">\n<ul>\n<li style="list-style-type: none;">\n<ul>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Writing automated functional and performance test cases from requirements, executing them in the lab, troubleshooting issues to identify their root cause and reporting results</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Troubleshooting system and network problems and diagnosing and solving hardware or software faults</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Reproducing issues in the lab and testing fixes and workarounds</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Writing procedural documentation and event reports</span></li>\n</ul>\n</li>\n</ul>\n</li>\n</ul>\n<p>\xa0</p>\n<p><strong>Qualifications and Skills Required</strong></p>\n<ul>\n<li style="list-style-type: none;">\n<ul>\n<li style="list-style-type: none;">\n<ul>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Academic and ideally professional experience writing software and/or scripts</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Experience with Python, BASH, and REST</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Knowledge of some of the following areas and drive to learn more: L2 to L4 of TCP/IP, traffic generators, hardware interfaces, chipset based solutions.</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">GIT, bug tracking software and documentation</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Excellent problem solving and troubleshooting skills</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Ability to write clear procedures and reports</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Ability to perform under pressure in a deadline driven environment</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Excellent team player with a high level of self-motivation and initiative</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Excellent communication skills, both verbal and written (multilingualism is an asset)</span></li>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">B Sc. in Computer Science, Software Engineering, B.Elec.Eng. completed</span></li>\n</ul>\n</li>\n</ul>\n</li>\n</ul>\n<p>\xa0</p>\n<p><b>Good to have</b></p>\n<ul>\n<li aria-level="1" style="font-weight: 400;"><span style="font-weight: 400;">Networking certifications</span></li>\n</ul>\n</div>\n</div>\n</div>\n</div>\n<div class="so-panel widget widget_sow-editor panel-last-child" data-index="7" id="panel-18-2-0-1">\n<div class="so-widget-sow-editor so-widget-sow-editor-base">\n<div class="siteorigin-widget-tinymce textwidget">\n<p>\xa0</p>\n<p><strong>Additional Information</strong></p>\n<p>Type: Full-time</p>\n<p>Location: Montreal, QC, Canada</p>\n<hr/>\n<p>For more\xa0information, or to submit your resumé, please\xa0e-mail\xa0NoviFlow at\xa0<a href="https://mail.google.com/mail/?view=cm&amp;fs=1&amp;tf=1&amp;to=careers@noviflow.com">careers@noviflow.com</a></p>\n</div>\n</div>\n</div>\n</div>\n</div>\n</div>\n</div>\n</div>\n</div>\n</div>\n</div></div>''dt',
 'em',
 'fieldset',
 'font',
 'footer',
 'form',
 'g',
 'h1',
 'h2',
 'h3',
 'h4',
 'h5',
 'h6',
...
 'thead',
 'time',
 'tr',
 'u',
 'ul'}
Output is truncated. View as a scrollable element or open in a text editor. Adjust cell output settings...
{'Your role as In-person Network Coordinator': " Reporting to the EAP Manager, the In-person Network Coordinator plays a vital role in enhancing our provider network's reach and effectiveness, as well as ensuring comprehensive support to our external mental health providers. What you’ll be doing:", 'What you’ll be doing:': " Execution of in-person provider recruitment, such as candidate sourcing, inbox management and contract drafting, in close collaboration with the Talent Acquisition SpecialistSupport manager in tracking and planning network coverage across CanadaProvide seamless onboarding and off-boarding experience, keeping internal management tools up-to-date and offering personalized coaching to providers to drive completion of HR and IT requirementsProvide and maintain a great ongoing provider experience, with responsive support and answer inquiries from providers, addressing their questions and concerns promptly Execution of in-person provider recruitment, such as candidate sourcing, inbox management and contract drafting, in close collaboration with the Talent Acquisition Specialist Support manager in tracking and planning network coverage across Canada Provide seamless onboarding and off-boarding experience, keeping internal management tools up-to-date and offering personalized coaching to providers to drive completion of HR and IT requirements Provide and maintain a great ongoing provider experience, with responsive support and answer inquiries from providers, addressing their questions and concerns promptly We'd love to hear from you if you have:", "We'd love to hear from you if you have:": ' Experience in network coordination, provider relations, administrative support or a related fieldStrong communication skills in French and English, both written and verbal, with the ability to effectively interact with providers and internal teamsDetail-oriented and able to maintain accurate records and documentationProficiency in Google Sheets or ExcelYou have a problem-solving mindset and are proactive in bringing your solutions to lifeAbility to work independently and as part of a team, collaborating effectively to achieve common goalsExperience in healthcare or a similar industry is a plus Experience in network coordination, provider relations, administrative support or a related field Strong communication skills in French and English, both written and verbal, with the ability to effectively interact with providers and internal teams Detail-oriented and able to maintain accurate records and documentation Proficiency in Google Sheets or Excel You have a problem-solving mindset and are proactive in bringing your solutions to life Ability to work independently and as part of a team, collaborating effectively to achieve common goals Experience in healthcare or a similar industry is a plus At Dialogue, your well-being is our priority', 'At Dialogue, your well-being is our priority': " Taking care of others also means taking care of our team. We’ve got you covered! Taking care of others also means taking care of our team. We’ve got you covered! A fully funded benefits plan, includinga wellness reimbursement programUnlimited access to a variety of Dialogue's programs for you and your immediate family4 weeks of vacation, 9 wellness days and 1 paid volunteer dayA flexible schedule and a hybrid work approachAn allocated budget for continuous trainingShort and long-term incentive plans, including restricted stock units (RSUs)An optional parental benefits program A fully funded benefits plan, includinga wellness reimbursement program A fully funded benefits plan, including a wellness reimbursement program Unlimited access to a variety of Dialogue's programs for you and your immediate family Unlimited access to a variety of Dialogue's programs for you and your immediate family 4 weeks of vacation, 9 wellness days and 1 paid volunteer day 4 weeks of vacation, 9 wellness days and 1 paid volunteer day A flexible schedule and a hybrid work approach A flexible schedule and a hybrid work approach An allocated budget for continuous training An allocated budget for continuous training Short and long-term incentive plans, including restricted stock units (RSUs) Short and long-term incentive plans, including restricted stock units (RSUs) An optional parental benefits program An optional parental benefits program About Dialogue", 'About Dialogue': ' Dialogue is the #1 virtual care provider in Canada. By developingour Integrated Health Platform🅫, we provide exceptional online health and wellness programs (primary care, mental health, iCBT, EAP, and wellness) to organizations that want to improve the wellness of their employees and families. Dialogue is the #1 virtual care provider in Canada. By developing   our Integrated Health Platform🅫, we provide exceptional online health and wellness programs (primary care, mental health, iCBT, EAP, and wellness) to organizations that want to improve the wellness of their employees and families. When it comes to our work, we set the bar high. Together, we’re transforming health and helping millions improve their well-being. We’re firm believers that great people don’t settle on: When it comes to our work, we set the bar high. Together, we’re transforming health and helping millions improve their well-being. We’re firm believers that great people don’t settle on: Impact Impact Community Community Growth Growth Excellence Excellence Feel like you can make a difference? Good news, we saved you a seat! Feel like you can make a difference? Good news, we saved you a seat!'}
li	p	ul	b	br	section	h3	h2	h4	h1	h5	h6
0	Driving a culture of achievement to exceed rev...	At Vidyard, we make life easier for sellers, m...	Driving a culture of achievement to exceed rev...	None	None	None	None	About the Role	None	None	None	None
1	Setting clear expectations around revenue gene...	Vidyard is looking for a Director Sales - Ent...	~ 5 years of proven sales leadership experienc...	None	None	None	None	About the Team	None	None	None	None
2	Working with marketing for in-market events an...	The ideal candidate for this role should have ...	Salesforce, Salesloft, Zoominfo & SalesNavigator	None	None	None	None	What You’ll Work On	None	None	None	None
3	Coaching team in effective sales strategies, p...	This is a remote role open to candidates locat...	Competitive pay Comprehensive, flexible bene...	None	None	None	None	What You’ll Bring to this Role and Your New Team:	None	None	None	None
4	Cultivating and maintaining strong relationshi...	Our Enterprise New Business team is a small ...	None	None	None	None	None	Our Sales Tech Stack	None	None	None	None
...	...	...	...	...	...	...	...	...	...	...	...	...
98821	Help	None	None	None	None	None	None	None	None	None	None	None
98822	Support \n \n Contact Support \n Customer Zone...	None	None	None	None	None	None	None	None	None	None	None
98823	Contact Support	None	None	None	None	None	None	None	None	None	None	None
98824	Customer Zone	None	None	None	None	None	None	None	None	None	None	None
98825	Help	None	None	None	None	None	None	None	None	None	None	None
98826 rows × 12 columns

li         90851
p          40688
ul         13744
b          11216
br             0
section     5057
h3          4319
h2          2102
h4           673
h1           340
h5           141
h6            22
dtype: int64
array(['#LI-RB1', 'Ready To Join Us?', 'Engineering Technician',
       'Industrial Hygienist',
       'Demo Artist & Trainer - Generalist (Job Req #2024-010)',
       'Senior FPGA Designer/Architect (Job Req #2024-008)', 'Summary:',
       'A day in the life might look like:',
       'You may be a fit for this role if:', 'Maintenance Technician',
       'Business Development Manager - Graphics, Americas (Job Req #2024-002)',
       'Customer Success Manager',
       'Regional Sales Manager, Southeast Asia (Job Req #2024-003)',
       'Manufacturing Training Specialist (Job Req #2024-001)',
       'Quality Control Chemist',
       'Senior Product Verification Specialist (4-Month Contract) (Job Req #2024-006)',
       'POSITION OVERVIEW', 'QUALIFICATIONS', 'COMPANY OVERVIEW',
       'Qualifications', 'Company Overview', 'Position Overview',
       'SOC Analyst I (Temporary Full-time: 12-month Contract)',
       'Quality Assurance Associate, Process Development',
       'Production Technician', 'PD Student Scientist',
       'Global Proposals Manager (Temporary Full-time: 14 month Contract)',
       'Global Proposals Manager (Temporary Full-time: 14-month Contract)',
       'Stagiaire - Génie électrique - Liaison - H24', 'Commis PDI',
       'Business Analyst (Oracle Fusion Cloud)', 'What your team does:',
       'Who you are:', 'What you’ll work on:', 'What you may have:',
       'Serious bonus points if you have:', 'What you should have:',
       'Serious bonus points if you:', '#LI-LI1',
       'Manager, Manufacturing Quality Systems',
...
       'Business Development Representative', 'Senior Account Executive',
       'Registered Dietitian', 'Registered Social Worker/Psychotherapist',
       'Registered Clinical Psychologist', 'Customer Success Specialist',
       'Full Time Customer Support Representative - Ottawa West',
       'Client Operations Assistant'], dtype=object)
Output is truncated. View as a scrollable element or open in a text editor. Adjust cell output settings...

# Test Automation Software Developer

**Job ID: 2023070104**

**Job Description**

You will design and develop an automated test framework designed for application layer and embedded software for bleeding-edge networking technologies, including ultra-fast network processors (up to 12.8Tbps). The products bring together Open Systems, Network Virtualization, and fully Programmable Network Logic to meet the needs of Data Centers, Network Service Providers, and researchers in Software Defined Networking technology.

**Responsibilities**

- Writing automated functional and performance test cases from requirements, executing them in the lab, troubleshooting issues to identify their root cause, and reporting results.
- Troubleshooting system and network problems and diagnosing and solving hardware or software faults.
- Reproducing issues in the lab and testing fixes and workarounds.
- Writing procedural documentation and event reports.

**Qualifications and Skills Required**

- Academic and ideally professional experience writing software and/or scripts.
- Experience with Python, BASH, and REST.
- Knowledge of some of the following areas and a drive to learn more: L2 to L4 of TCP/IP, traffic generators, hardware interfaces, chipset-based solutions.
- GIT, bug tracking software, and documentation.
- Excellent problem-solving and troubleshooting skills.
- Ability to write clear procedures and reports.
- Ability to perform under pressure in a deadline-driven environment.
- Excellent team player with a high level of self-motivation and initiative.
- Excellent communication skills, both verbal and written (multilingualism is an asset).
- B Sc. in Computer Science, Software Engineering, B.Elec.Eng. completed.

**Good to have**

- Networking certifications.

**Additional Information**

- Type: Full-time
- Location: Montreal, QC, Canada

---

For more information or to submit your resumé, please email NoviFlow at [careers@noviflow.com](https://mail.google.com/mail/?view=cm&fs=1&tf=1&to=careers@noviflow.com).

In [53]:
unique_tags = set()

for description in data['description']:
    if not isinstance(description, str):
        continue

    soup = BeautifulSoup(description, 'html.parser')
    for tag in soup.find_all(True):
        unique_tags.add(tag.name)

unique_tags


{'a',
 'abbr',
 'article',
 'aside',
 'b',
 'blockquote',
 'br',
 'button',
 'circle',
 'dd',
 'div',
 'dl',
 'dt',
 'em',
 'fieldset',
 'font',
 'footer',
 'form',
 'g',
 'h1',
 'h2',
 'h3',
 'h4',
 'h5',
 'h6',
 'header',
 'hr',
 'i',
 'img',
 'input',
 'label',
 'legend',
 'li',
 'main',
 'nav',
 'ol',
 'option',
 'p',
 'path',
 's',
 'section',
 'select',
 'small',
 'span',
 'strong',
 'style',
 'sub',
 'sup',
 'svg',
 'table',
 'tbody',
 'td',
 'textarea',
 'th',
 'thead',
 'time',
 'tr',
 'u',
 'ul'}

In [54]:
def extract_heading_content_pairs(html_content):
    soup = BeautifulSoup(html_content, 'html.parser')
    
    heading_tags = {'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'strong'}
    heading_content_pairs = {}
    current_heading = None

    for element in soup.find_all(True): 
        if element.name in heading_tags:
            current_heading = element.get_text(strip=True)
            heading_content_pairs[current_heading] = ''
        elif current_heading:
            heading_content_pairs[current_heading] += ' ' + element.get_text(strip=True)

    return heading_content_pairs

description = data.iloc[123]['description']
heading_content_pairs = extract_heading_content_pairs(description)
print(heading_content_pairs)


{'Your role as In-person Network Coordinator': " Reporting to the EAP Manager, the In-person Network Coordinator plays a vital role in enhancing our provider network's reach and effectiveness, as well as ensuring comprehensive support to our external mental health providers. What you’ll be doing:", 'What you’ll be doing:': " Execution of in-person provider recruitment, such as candidate sourcing, inbox management and contract drafting, in close collaboration with the Talent Acquisition SpecialistSupport manager in tracking and planning network coverage across CanadaProvide seamless onboarding and off-boarding experience, keeping internal management tools up-to-date and offering personalized coaching to providers to drive completion of HR and IT requirementsProvide and maintain a great ongoing provider experience, with responsive support and answer inquiries from providers, addressing their questions and concerns promptly Execution of in-person provider recruitment, such as candidate so

In [55]:
def parse_html_based_on_tags(html_content, tags_to_extract):
    if isinstance(html_content, str):
        soup = BeautifulSoup(html_content, 'html.parser')


        parsed_data = {}

        for tag_name in tags_to_extract:
            tag_data = []
            for tag in soup.find_all(tag_name):
                tag_text = tag.get_text(separator=' ').strip()
                if tag_text:
                    tag_data.append(tag_text)
            parsed_data[tag_name] = tag_data

        max_length = max(len(data) for data in parsed_data.values())

        for tag_name, tag_data in parsed_data.items():
            if len(tag_data) < max_length:
                parsed_data[tag_name] += [None] * (max_length - len(tag_data))

        df = pd.DataFrame(parsed_data)

        return df
    else:

        return pd.DataFrame()


tags_to_extract = unique_tags_df['Tag'].tolist()


combined_parsed_df = pd.DataFrame()

for description in data['description']:
    parsed_df = parse_html_based_on_tags(description, tags_to_extract)

    combined_parsed_df = pd.concat([combined_parsed_df, parsed_df], ignore_index=True)
combined_parsed_df

li  \
0      Driving a culture of achievement to exceed rev...   
1      Setting clear expectations around revenue gene...   
2      Working with marketing for in-market events an...   
3      Coaching team in effective sales strategies, p...   
4      Cultivating and maintaining strong relationshi...   
...                                                  ...   
98821                                               Help   
98822  Support \n \n Contact Support \n Customer Zone...   
98823                                    Contact Support   
98824                                      Customer Zone   
98825                                               Help   

                                                       p  \
0      At Vidyard, we make life easier for sellers, m...   
1      Vidyard is looking for a  Director Sales - Ent...   
2      The ideal candidate for this role should have ...   
3      This is a remote role open to candidates locat...   
4      Our  Enterprise New Business team  is a small ...   
...                                                  ...   
98821                                               None   
98822                                               None   
98823                                               None   
98824                                               None   
98825                                               None   

                                                      ul     b    br section  \
0      Driving a culture of achievement to exceed rev...  None  None    None   
1      ~ 5 years of proven sales leadership experienc...  None  None    None   
2       Salesforce, Salesloft, Zoominfo & SalesNavigator  None  None    None   
3      Competitive pay   Comprehensive, flexible bene...  None  None    None   
4                                                   None  None  None    None   
...                                                  ...   ...   ...     ...   
98821                                               None  None  None    None   
98822                                               None  None  None    None   
98823                                               None  None  None    None   
98824                                               None  None  None    None   
98825                                               None  None  None    None   

         h3                                                 h2    h4    h1  \
0      None                                     About the Role  None  None   
1      None                                     About the Team  None  None   
2      None                                What You’ll Work On  None  None   
3      None  What You’ll Bring to this Role and Your New Team:  None  None   
4      None                               Our Sales Tech Stack  None  None   
...     ...                                                ...   ...   ...   
98821  None                                               None  None  None   
98822  None                                               None  None  None   
98823  None                                               None  None  None   
98824  None                                               None  None  None   
98825  None                                               None  None  None   

         h5    h6  
0      None  None  
1      None  None  
2      None  None  
3      None  None  
4      None  None  
...     ...   ...  
98821  None  None  
98822  None  None  
98823  None  None  
98824  None  None  
98825  None  None  

[98826 rows x 12 columns]

In [56]:
combined_parsed_df.count()

li         90851
p          40688
ul         13744
b          11216
br             0
section     5057
h3          4319
h2          2102
h4           673
h1           340
h5           141
h6            22
dtype: int64

In [57]:
unique_h1_values = combined_parsed_df[combined_parsed_df['h1'].notnull()]['h1'].unique()
unique_h1_values

array(['#LI-RB1', 'Ready To Join Us?', 'Engineering Technician',
       'Industrial Hygienist',
       'Demo Artist & Trainer - Generalist (Job Req #2024-010)',
       'Senior FPGA Designer/Architect (Job Req #2024-008)', 'Summary:',
       'A day in the life might look like:',
       'You may be a fit for this role if:', 'Maintenance Technician',
       'Business Development Manager - Graphics, Americas (Job Req #2024-002)',
       'Customer Success Manager',
       'Regional Sales Manager, Southeast Asia (Job Req #2024-003)',
       'Manufacturing Training Specialist (Job Req #2024-001)',
       'Quality Control Chemist',
       'Senior Product Verification Specialist (4-Month Contract) (Job Req #2024-006)',
       'POSITION OVERVIEW', 'QUALIFICATIONS', 'COMPANY OVERVIEW',
       'Qualifications', 'Company Overview', 'Position Overview',
       'SOC Analyst I (Temporary Full-time: 12-month Contract)',
       'Quality Assurance Associate, Process Development',
       'Production Techn

In [61]:
 > BERT
 > LLM  - PFT parameter efficient fine tune. is a catigory.
 two methods : P tuning & lora(low rank adaptation)
 4 bin - condisato - GPU moemory
 8GB ( nvidia) 3050  
 M series mac. 
 

SyntaxError: invalid syntax (3791788704.py, line 1)

In [ ]:
# LLama - 7B. 
# 32B or GPT4 as teacher to 7B
manualy label only 100 datset ,.. GPT4 to use the job.

